keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,142 @@
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Pod Successful Completion Cases](#pod-successful-completion-cases)
+    - [Job Parallelism Equal To 1](#job-parallelism-equal-to-1)
+    - [Job Parallelism Greater Than 1](#job-parallelism-greater-than-1)
+  - [Pod Failure Cases](#pod-failure-cases)
+    - [RestartPolicy of Pods as Never](#restartpolicy-of-pods-as-never)
+    - [RestartPolicy of Pods as OnFailure](#restartpolicy-of-pods-as-onfailure)
+- [API](#api)
+    - [ReclaimedPodSetPods](#reclaimedpodsetpods)
+- [Implementation](#implementation)
+- [Testing Plan](#testing-plan)
+  - [Unit Tests](#unit-tests)
+  - [Integration tests](#integration-tests)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This proposal allows Kueue to reclaim the resources of a successfully completed Pod of a Workload even before the whole Workload completes its execution. Freeing the unused resources of a Workload earlier may allow more pending Workloads to be admitted. This will allow Kueue to utilize the existing resources of the cluster much more efficiently.
+
+In the remaining of this document, *Pods of a Workload* means the same as *Pods of a Job* and vice-versa.
+
+
+## Motivation
+
+Currently, the resources owned by a Job are reclaimed by Kueue only when the whole Job finishes. For Jobs with multiple Pods, the resources will be reclaimed only after the last Pod of the Job finishes. This is not efficient as the Pods of parallel Job may have laggards consuming the unused resources of the Jobs those are currently executing.
+
+### Goals
+
+- Utilize the unused resources of the successfully completed Pods of a running Job.
+
+### Non-Goals
+
+- Preempting a Workload or Pods of a Workload to free resources for a pending Workload.
+- Partially admitting a Workload.
+
+## Proposal
+
+Reclaiming the resources of the succeeded Pods of a running Job as soon as the Pod completes its execution.
+
+We propose to add a new field `.status.ReclaimedPodSetPods` to the Workload API. Workload controller will be responsible for updating the field `.status.ReclaimedPodSetPods` according to the status of the Workload's Pods.
+
+`.status.ReclaimedPodSetPods` is a list that holds the count of successfully completed Pods belonging to a PodSet whose resources have been reclaimed by Kueue.
+
+Now we will look at the Pod failure cases and how the reclaiming of resources by Kueue will happen.
+
+### Pod Successful Completion Cases
+
+#### Job Parallelism Equal To 1
+
+By default, the `.spec.parallelism` of Job is equal to `1`.  In this case, if the Pod completes with successful execution, the whole Job execution can be considered as a success. Hence, the resource associated with the Job will be reclaimed by Kueue after the Job completion. This functionality is present today as well and no change will be required.
+
+#### Job Parallelism Greater Than 1
```

Comment:
Note that parallelism is not the same as completions.

Did you analyze the case where parallelism=N and completions=M, where N<M?

If a pod finishes successfully, it doesn't necessarily mean that the job controller won't create another one for it.

Reply:
The remaining completions of a job will reduce by one when its pod will finish successfully. I think, we should reclaim the resources of a successful pod when the job's parallelism is greater or equal to the remaining job completions value. In this way, we can also generalize this approach to applicable in any scenario for reclaiming the resources of a successful pod of a job .

@alculquicondor @kerthcet  let me know your comments on this. Just don't want to directly jump on writing the approach.

Reply:
let's say parallelism=3 and completions=5.
When 1 pod finishes (let's sayindex 0), there should still be 3 running pods (indexes 1, 2, 3)
When 3 pods finish, then there would only be 2 running pods (indexes 3, 4).

Is this what you mean?

Reply:
What you have explained is correct. That's what I meant. 

I shouldn't have said Job's parallelism greater than it's completions. Rather, as we see from the above scenario you have explained, when indexes 3 & 4 remain(rather 2, 3 & 4 remain, consider 0 &1 completed for simplicity), now remaining completions of Job is not greater than the possible Job's parallelism. After this point in the Job's lifetime, Kueue can start to reclaim the resources of the Pod which completes successfully.